## Hello!

This is an attempt to predict the winner of an NHL game given the NHL database given by Kaggle [here](https://www.kaggle.com/martinellis/nhl-game-data). After playing around with the data a lot, I found that home ice advantage played a role in the outcome of the game (home wins about 55% of the time). I looked into the head-to-head of each team vs. each other team and the results are found at this [imgur link](https://imgur.com/a/o316axx). After completing that, I wanted to find some feature that played a role in a team winning the game or not. Firstly, I mapped out where on the ice each team scored from the most, and where each team got scored on from the most - but this was not useful enough. Then, in this file, I checked the win rate if a team scores first - which was found to be useful and is explored below.

In [1]:
import pandas as pd
#Read in csv's.
#Need to replace team_id's because this database includes 33 teams (Phoenix and Atlanta had name changes to Arizona and Winnipeg)
#which the database defaulted to calling team_id 52,53,54. I replace them to have team_id's go from 1-31

iter_gp_csv = pd.read_csv(r'C:\Users\SimoS\Documents\NHL\game_plays.csv', iterator = True, chunksize = 1000)
game_plays_df = pd.concat([chunk[(chunk.event == 'Goal')][['game_id','team_id_for', 'team_id_against', 'event']] for chunk in iter_gp_csv])
game_plays_df = game_plays_df.replace([52,53,54], [11,27,31])

#Only read in Regular Season games. Again, replace necessary team id's
iter_game_csv = pd.read_csv(r'C:\Users\SimoS\Documents\NHL\game.csv', iterator = True, chunksize = 1000)
game_df = pd.concat([chunk[(chunk.type == 'R')] for chunk in iter_game_csv])
game_df = game_df.replace([52,53,54], [11,27,31])

iter_gts_csv = pd.read_csv(r'C:\Users\SimoS\Documents\NHL\game_teams_stats.csv', iterator = True, chunksize = 1000)
game_teams_stats_df = pd.concat([chunk for chunk in iter_gts_csv])
game_teams_stats_df = game_teams_stats_df.replace([52,53,54], [11,27,31])

iter_team_csv = pd.read_csv(r'C:\Users\SimoS\Documents\NHL\team_info.csv', iterator = True, chunksize = 1000)
team_df = pd.concat([chunk for chunk in iter_team_csv])

In [2]:
#Let's look at all the dataframes
game_plays_df.head(5)

game_id  team_id_for  team_id_against event
15   2011030221          1.0              4.0  Goal
133  2011030221          4.0              1.0  Goal
141  2011030221          4.0              1.0  Goal
161  2011030221          1.0              4.0  Goal
236  2011030221          4.0              1.0  Goal

In [3]:
game_df.head(5)

game_id    season type   date_time         date_time_GMT  \
705  2013020674  20132014    R  2014-01-11  2014-01-11T03:00:00Z   
706  2013020177  20132014    R  2013-10-29  2013-10-29T23:00:00Z   
707  2012020225  20122013    R  2013-02-19  2013-02-19T03:00:00Z   
708  2012020577  20122013    R  2013-04-08  2013-04-08T23:00:00Z   
709  2013021085  20132014    R  2014-03-25  2014-03-25T23:30:00Z   

     away_team_id  home_team_id  away_goals  home_goals       outcome  \
705            19            23           1           2  home win REG   
706            24             4           3           2  away win REG   
707            29            24           2           3  home win REG   
708            12             6           2           6  home win REG   
709            17            29           2           4  home win REG   

    home_rink_side_start               venue           venue_link  \
705                right        Rogers Arena  /api/v1/venues/null   
706                right  Wells Fargo Center  /api/v1/venues/null   
707                 left        Honda Center  /api/v1/venues/null   
708                 left           TD Garden  /api/v1/venues/null   
709                 left    Nationwide Arena  /api/v1/venues/null   

      venue_time_zone_id  venue_time_zone_offset venue_time_zone_tz  
705    America/Vancouver                      -7                PDT  
706     America/New_York                      -4                EDT  
707  America/Los_Angeles                      -7                PDT  
708     America/New_York                      -4                EDT  
709     America/New_York                      -4                EDT

In [4]:
game_teams_stats_df.head(5)

game_id  team_id   HoA    won settled_in        head_coach  goals  \
0  2011030221        1  away  False         OT      Peter DeBoer      3   
1  2011030221        4  home   True         OT  Peter Laviolette      4   
2  2011030222        1  away   True        REG      Peter DeBoer      4   
3  2011030222        4  home  False        REG  Peter Laviolette      1   
4  2011030223        4  away  False         OT  Peter Laviolette      3   

   shots  hits  pim  powerPlayOpportunities  powerPlayGoals  \
0     26    31   12                       3               1   
1     36    27    6                       6               1   
2     35    32   12                       4               0   
3     20    24   32                       5               0   
4     28    28    4                       5               1   

   faceOffWinPercentage  giveaways  takeaways  
0                  44.9          6          7  
1                  55.1         13          4  
2                  50.9          8          7  
3                  49.1          9          6  
4                  50.8          2          1

In [5]:
team_df.head(5)

team_id  franchiseId     shortName   teamName abbreviation  \
0        1           23    New Jersey     Devils          NJD   
1        4           16  Philadelphia     Flyers          PHI   
2       26           14   Los Angeles      Kings          LAK   
3       14           31     Tampa Bay  Lightning          TBL   
4        6            6        Boston     Bruins          BOS   

               link  
0   /api/v1/teams/1  
1   /api/v1/teams/4  
2  /api/v1/teams/26  
3  /api/v1/teams/14  
4   /api/v1/teams/6

In [6]:
#These are some simple functions that will be used later in the file.

"""
Return a complete, cleaned teamList
"""
def teamList():
    teamList = team_df[['team_id','shortName','teamName']][(team_df.team_id != 11) & (team_df.team_id != 27)].sort_values("shortName")
    teamList = teamList.replace([52,53,54],[11,27,31])
    return teamList
"""
Get the team name given teamID
"""
def getTeamName(userTeamID):
    x = teamList()
    teamName = x[(x.team_id == userTeamID)]
    teamName = teamName[['teamName']]
    return teamName.iloc[0][0]



In [7]:
#Get the first goal of each game
goals_df = game_plays_df
goals_df = goals_df.drop_duplicates("game_id")
score_first_df = goals_df.merge(game_teams_stats_df, on='game_id')
score_first_df = score_first_df[score_first_df.team_id_for == score_first_df.team_id]
score_first_df.head(10)

game_id  team_id_for  team_id_against event  team_id   HoA    won  \
0   2011030221          1.0              4.0  Goal        1  away  False   
3   2011030222          4.0              1.0  Goal        4  home  False   
4   2011030223          4.0              1.0  Goal        4  away  False   
6   2011030224          4.0              1.0  Goal        4  away  False   
9   2011030225          4.0              1.0  Goal        4  home  False   
10  2011030411         26.0              1.0  Goal       26  away   True   
12  2011030412         26.0              1.0  Goal       26  away   True   
15  2011030413         26.0              1.0  Goal       26  home   True   
16  2011030414          1.0             26.0  Goal        1  away   True   
19  2011030415          1.0             26.0  Goal        1  home   True   

   settled_in        head_coach  goals  shots  hits  pim  \
0          OT      Peter DeBoer      3     26    31   12   
3         REG  Peter Laviolette      1     20    24   32   
4          OT  Peter Laviolette      3     28    28    4   
6         REG  Peter Laviolette      2     22    23   10   
9         REG  Peter Laviolette      1     28    38    8   
10         OT     Darryl Sutter      2     25    35    4   
12         OT     Darryl Sutter      2     32    32    8   
15        REG     Darryl Sutter      4     21    55   12   
16        REG      Peter DeBoer      3     24    43    8   
19        REG      Peter DeBoer      2     19    34    6   

    powerPlayOpportunities  powerPlayGoals  faceOffWinPercentage  giveaways  \
0                        3               1                  44.9          6   
3                        5               0                  49.1          9   
4                        5               1                  50.8          2   
6                        2               1                  62.5         20   
9                        1               0                  45.0         14   
10                       1               0                  55.4          6   
12                       2               0                  53.5          1   
15                       2               2                  56.6         17   
16                       3               0                  31.0         13   
19                       2               1                  52.8          4   

    takeaways  
0           7  
3           6  
4           1  
6           5  
9           7  
10          4  
12          3  
15          5  
16          2  
19          3

In [8]:
#Get a cleaned team list
team_list = teamList()
team_df_modified = team_list[['team_id', 'shortName']]
team_df_modified

team_id     shortName
17       24       Anaheim
31       27       Arizona
4         6        Boston
22        7       Buffalo
23       20       Calgary
30       12      Carolina
11       16       Chicago
24       21      Colorado
26       29      Columbus
25       25        Dallas
7        17       Detroit
28       22      Edmonton
21       13       Florida
2        26   Los Angeles
14       30     Minnesota
13        8      Montreal
19        2  NY Islanders
5         3    NY Rangers
9        18     Nashville
0         1    New Jersey
12        9        Ottawa
1         4  Philadelphia
6         5    Pittsburgh
8        28      San Jose
16       19      St Louis
3        14     Tampa Bay
20       10       Toronto
10       23     Vancouver
29       31         Vegas
15       15    Washington
27       11      Winnipeg

In [9]:
#See how many times each team scored first in the last 3 years (246 games). Merging with game_df eliminates playoff games.
score_first = pd.merge(score_first_df,game_df, on='game_id')[['game_id','team_id', 'HoA', 'won']]
score_first = score_first[((score_first.game_id.astype(str).str.startswith('2016'))
                            | (score_first.game_id.astype(str).str.startswith('2017')) 
                            | (score_first.game_id.astype(str).str.startswith('2018')))]
score_first.head(15)

game_id  team_id   HoA    won
3125  2016020045       16  home   True
3126  2017020812        7  home  False
3129  2017020586       24  home   True
3130  2016020610       15  away   True
3132  2017020240       23  away  False
3133  2017020624       21  home   True
3135  2017020221       11  away   True
3136  2016020298       26  home   True
3137  2017020731       27  home   True
3138  2016020205       11  away   True
3139  2016020246        4  home   True
3140  2016020107        1  home  False
3141  2016021020        5  away  False
3142  2017020986       14  away   True
3143  2017021124       12  home  False

In [10]:
score_first_count = score_first['team_id'].value_counts().reset_index()
score_first_count = score_first_count.rename(columns={'team_id':'score_first','index':'team_id'})
score_first_count = score_first_count.merge(team_df_modified, on='team_id')
score_first_count 

team_id  score_first     shortName
0         5          140    Pittsburgh
1        16          138       Chicago
2        19          137      St Louis
3        29          137      Columbus
4        15          136    Washington
5        10          135       Toronto
6        28          129      San Jose
7        27          128       Arizona
8         6          128        Boston
9        11          127      Winnipeg
10       18          126     Nashville
11       13          126       Florida
12        1          125    New Jersey
13        2          125  NY Islanders
14        8          124      Montreal
15       12          123      Carolina
16       24          123       Anaheim
17       25          122        Dallas
18       14          121     Tampa Bay
19        3          118    NY Rangers
20        7          116       Buffalo
21       22          116      Edmonton
22       17          114       Detroit
23       30          114     Minnesota
24       20          112       Calgary
25       21          110      Colorado
26        9          110        Ottawa
27        4          100  Philadelphia
28       23           93     Vancouver
29       26           92   Los Angeles
30       31           83         Vegas

In [11]:
home_wins_score_first = score_first[score_first.HoA == 'home']['won'].value_counts()
home_wins_score_first

True     1387
False     547
Name: won, dtype: int64

In [12]:
1387/(1387+547)

0.7171664943123061

Win 71.72% when scoring first at home. Let's look at scoring first at an away game

In [13]:
away_wins_score_first = score_first[score_first.HoA == 'away']['won'].value_counts()
away_wins_score_first


True     1120
False     674
Name: won, dtype: int64

In [14]:
1120/(1120+674)

0.6243032329988851

Win 62.24% when scoring first away. This is a big difference from the 70% at home. Home ice advantage is significant. Using what we now know (home ice advantage, scoring first, and how often a team scores first), let's see what training our model can do.

### This is currently being worked on and updated - this is just showing how to train a model

In [15]:
#Create train set (2016-2017, 2017-2018, 2018-2019 seasons)
train = game_df[['game_id', 'home_team_id', 'away_team_id', 'outcome']]
train = train[((train.game_id.astype(str).str.startswith('2016')) | (train.game_id.astype(str).str.startswith('2017')) | (train.game_id.astype(str).str.startswith('2018')))]

In [16]:
train.loc[train.outcome.str.contains('home'), 'outcome'] = 'home_win'
train.loc[train.outcome.str.contains('away',na=False), 'outcome'] = 'away_win'

In [17]:
train

game_id  home_team_id  away_team_id   outcome
3885   2016020045            16             4  home_win
3886   2017020812             7            24  away_win
3889   2017020586            24            20  home_win
3890   2016020610             8            15  away_win
3892   2017020240            24            23  home_win
3893   2017020624            21            29  home_win
3895   2017020221            25            11  away_win
3896   2016020298            26             2  home_win
3897   2017020731            27             2  home_win
3898   2016020205            27            11  away_win
3899   2016020246             4            11  home_win
3900   2016020107             1            16  away_win
3901   2016021020            20             5  home_win
3902   2017020986            25            14  away_win
3903   2017021124            12            22  away_win
3905   2017020146            30             2  home_win
3906   2016020716             2            29  home_win
3907   2017020001            11            10  away_win
3908   2016020933            28            10  home_win
3911   2017020482             8             1  home_win
3912   2016021163            11             9  home_win
3914   2017020706            15             8  away_win
3916   2017021076            20            22  home_win
3918   2016020264             8            10  home_win
3919   2017021046            25            24  home_win
3922   2017020214             2            21  home_win
3923   2016020683            24            21  home_win
3924   2017020285            21            15  home_win
3925   2017020423            23            12  home_win
3927   2017020930            10             2  home_win
...           ...           ...           ...       ...
11317  2018021242             7             9  home_win
11318  2018021243            10            14  away_win
11319  2018021244            13             2  away_win
11320  2018021245             5            17  home_win
11321  2018021246            15             8  home_win
11322  2018021247            12             1  home_win
11323  2018021248            19             4  home_win
11324  2018021249            18            23  home_win
11325  2018021250            30             6  away_win
11326  2018021251            21            11  home_win
11327  2018021252            22            28  away_win
11328  2018021253            31            27  away_win
11329  2018021254             3            29  away_win
11330  2018021255            16            25  home_win
11331  2018021256            24            26  home_win
11332  2018021257             6            14  away_win
11333  2018021258            19            23  home_win
11334  2018021259             8            10  home_win
11335  2018021260             9            29  away_win
11336  2018021261            17             7  away_win
11337  2018021262            13             1  away_win
11338  2018021263             4            12  away_win
11339  2018021264             5             3  away_win
11340  2018021265            15             2  away_win
11341  2018021266            18            16  home_win
11342  2018021267            25            30  home_win
11343  2018021269            27            11  away_win
11344  2018021268            20            22  away_win
11345  2018021270            26            31  home_win
11346  2018021271            28            21  home_win

[3772 rows x 4 columns]

In [18]:
#Get head to head stats of the last 3 seasons
g = game_df[((game_df.type == "R") & ((game_df.game_id.astype(str).str.startswith('2016')) | (game_df.game_id.astype(str).str.startswith('2017')) 
                | (game_df.game_id.astype(str).str.startswith('2018'))))][['home_team_id', 'away_team_id', 'outcome']]

"""
This function is used to get the head_to_head stats of two teams.
"""
def train_head_to_head_aux(userTeam1, userTeam2):
    g_2 = g[(((g.home_team_id == userTeam2) &  (g.outcome.str.contains('home',na=False)))
             | ((g.away_team_id == userTeam2) & (g.outcome.str.contains('away',na=False))))]
    count = g_2['away_team_id'].value_counts()
    count2 = g_2['home_team_id'].value_counts()
    y = pd.concat([count, count2], axis = 1).reset_index()
    #Fix the NaN's
    y = y.fillna(0)
    userTeam2_name = getTeamName(userTeam2) + '_wins'
    y[userTeam2_name] = y['away_team_id'] + y['home_team_id']
    y[userTeam2_name] = pd.to_numeric(y[userTeam2_name], downcast='integer')
    y['away_team_id'] = pd.to_numeric(y['away_team_id'], downcast='integer')
    y = y.rename(columns={"away_team_id":"away_wins", "home_team_id":"home_wins","index":"team_id"})
    y = y[y.team_id == userTeam1]
    if y.empty == True:
        return 0
    return y.iloc[0][3]

### Need to populate training data with head to head stats:
Using our head_to_head function above, we can calculate a quotient in wins between each team. If Vegas has won 4 games against Chicago, and Chicago has won 2 against Vegas, it will give Vegas a 0.67 (4/6 = 2/3). 

In [19]:
home_team_list = train.home_team_id.to_list()
away_team_list = train.away_team_id.to_list()
count = 0
hth_list = []
for x in home_team_list:
    y = away_team_list[count]
    away = train_head_to_head_aux(y,x)
    home = train_head_to_head_aux(x, y)
    if away == 0:
        hth_list.append(1)
    else:
        hth_list.append(home / (home+away))
    count+=1


In [20]:
train['head_to_head'] = hth_list

In [21]:
train.head(10)

game_id  home_team_id  away_team_id   outcome  head_to_head
3885  2016020045            16             4  home_win      0.666667
3886  2017020812             7            24  away_win      0.333333
3889  2017020586            24            20  home_win      0.384615
3890  2016020610             8            15  away_win      0.666667
3892  2017020240            24            23  home_win      0.307692
3893  2017020624            21            29  home_win      0.666667
3895  2017020221            25            11  away_win      0.692308
3896  2016020298            26             2  home_win      0.666667
3897  2017020731            27             2  home_win      0.666667
3898  2016020205            27            11  away_win      0.777778

In [22]:
#Merge with the scoring first rates we calculated near the beginning
train = train.merge(score_first_count, left_on='home_team_id', right_on='team_id')
train['home_score_first'] = train['score_first'] / 246
train = train.drop(columns=['team_id','score_first','shortName'])

In [23]:
train

game_id  home_team_id  away_team_id   outcome  head_to_head  \
0     2016020045            16             4  home_win      0.666667   
1     2017021255            16            19  away_win      0.428571   
2     2016020178            16            25  home_win      0.384615   
3     2016020139            16            20  home_win      0.777778   
4     2016021067            16            21  home_win      0.500000   
5     2016020917            16            19  home_win      0.428571   
6     2017021040            16            12  away_win      0.666667   
7     2017020181            16             4  home_win      0.666667   
8     2016021152            16            29  home_win      0.500000   
9     2017020718            16             2  away_win      0.333333   
10    2017020821            16            20  away_win      0.777778   
11    2016020587            16            12  home_win      0.666667   
12    2017020054            16            30  away_win      0.384615   
13    2016020072            16            10  home_win      0.500000   
14    2017020681            16            17  away_win      0.333333   
15    2016020988            16            24  away_win      0.555556   
16    2016021012            16            30  home_win      0.384615   
17    2016020722            16            14  away_win      1.000000   
18    2017020927            16             9  home_win      0.166667   
19    2017021059            16             6  home_win      0.666667   
20    2016020474            16            28  home_win      0.666667   
21    2017020071            16            18  home_win      0.461538   
22    2017020729            16            14  away_win      1.000000   
23    2017021196            16            11  home_win      0.714286   
24    2016020374            16            11  away_win      0.714286   
25    2017020902            16            15  home_win      0.666667   
26    2016020617            16            17  home_win      0.333333   
27    2017020442            16             7  home_win      0.166667   
28    2017021174            16            28  away_win      0.666667   
29    2017021113            16            19  away_win      0.428571   
...          ...           ...           ...       ...           ...   
3742  2018020295             9             5  home_win      0.555556   
3743  2018020312             9            13  away_win      0.583333   
3744  2018020384             9             3  home_win      0.333333   
3745  2018020395             9            28  home_win      0.500000   
3746  2018020433             9             8  away_win      0.692308   
3747  2018020447             9             5  home_win      0.555556   
3748  2018020456             9             6  away_win      0.666667   
3749  2018020515             9            18  home_win      0.666667   
3750  2018020553             9            15  away_win      0.888889   
3751  2018020589             9            15  away_win      0.888889   
3752  2018020615             9            23  away_win      0.500000   
3753  2018020637             9            30  away_win      1.000000   
3754  2018020647             9            12  away_win      0.666667   
3755  2018020721             9            21  home_win      0.166667   
3756  2018020761             9            27  away_win      0.666667   
3757  2018020792             9            17  away_win      0.666667   
3758  2018020829             9            24  home_win      0.333333   
3759  2018020843             9            11  home_win      0.666667   
3760  2018020867             9            12  away_win      0.666667   
3761  2018020940             9            29  away_win      0.666667   
3762  2018020959             9            20  away_win      0.833333   
3763  2018020987             9            22  away_win      0.333333   
3764  2018021035             9             2  away_win      0.666667   
3765  2018021086             9            19  hom

In [24]:
train = train.merge(score_first_count, left_on='away_team_id', right_on='team_id')
train

game_id  home_team_id  away_team_id   outcome  head_to_head  \
0     2016020045            16             4  home_win      0.666667   
1     2017020181            16             4  home_win      0.666667   
2     2018021142            16             4  away_win      0.666667   
3     2017020539             7             4  home_win      0.666667   
4     2016020973             7             4  away_win      0.666667   
5     2016020613             7             4  home_win      0.666667   
6     2018020318             7             4  home_win      0.666667   
7     2018020444             7             4  away_win      0.666667   
8     2016020561            24             4  home_win      0.500000   
9     2017020029            24             4  away_win      0.500000   
10    2018020176            24             4  away_win      0.500000   
11    2016020164             8             4  home_win      0.555556   
12    2016020081             8             4  home_win      0.555556   
13    2017020965             8             4  away_win      0.555556   
14    2018020743             8             4  away_win      0.555556   
15    2018020934             8             4  home_win      0.555556   
16    2017021188            21             4  away_win      0.500000   
17    2016020444            21             4  away_win      0.500000   
18    2018020026            21             4  home_win      0.500000   
19    2017021183            25             4  home_win      0.500000   
20    2016020460            25             4  home_win      0.500000   
21    2018021235            25             4  home_win      0.500000   
22    2017020012            26             4  home_win      0.333333   
23    2016020016            26             4  away_win      0.333333   
24    2018020190            26             4  away_win      0.333333   
25    2017020850            27             4  away_win      0.500000   
26    2016020028            27             4  home_win      0.500000   
27    2018020214            27             4  away_win      0.500000   
28    2017020778             1             4  home_win      0.384615   
29    2016021180             1             4  home_win      0.384615   
...          ...           ...           ...       ...           ...   
3742  2016020685             6            16  away_win      0.333333   
3743  2017021048             6            16  home_win      0.333333   
3744  2018020863             6            16  home_win      0.333333   
3745  2016020433             3            16  away_win      0.666667   
3746  2017020615             3            16  away_win      0.666667   
3747  2018020728             3            16  home_win      0.666667   
3748  2017021239            19            16  away_win      0.571429   
3749  2016020562            19            16  home_win      0.571429   
3750  2017020093            19            16  home_win      0.571429   
3751  2016020467            19            16  away_win      0.571429   
3752  2016020195            19            16  away_win      0.571429   
3753  2018020024            19            16  away_win      0.571429   
3754  2018020155            19            16  home_win      0.571429   
3755  2017020328            14            16  home_win      0.000000   
3756  2016021124            14            16  home_win      0.000000   
3757  2018020341            14            16  home_win      0.000000   
3758  2017020300             5            16  away_win      1.000000   
3759  2016021137             5            16  away_win      1.000000   
3760  2018020650             5            16  away_win      1.000000   
3761  2017020874            31            16  home_win      0.000000   
3762  2017020135            31            16  home_win      0.000000   
3763  2018020438            31            16  home_win      0.000000   
3764  2016020281            22            16  home_win      0.444444   
3765  2016020827            22            16  awa

In [25]:
train['away_score_first'] = train['score_first'] / 246
train

game_id  home_team_id  away_team_id   outcome  head_to_head  \
0     2016020045            16             4  home_win      0.666667   
1     2017020181            16             4  home_win      0.666667   
2     2018021142            16             4  away_win      0.666667   
3     2017020539             7             4  home_win      0.666667   
4     2016020973             7             4  away_win      0.666667   
5     2016020613             7             4  home_win      0.666667   
6     2018020318             7             4  home_win      0.666667   
7     2018020444             7             4  away_win      0.666667   
8     2016020561            24             4  home_win      0.500000   
9     2017020029            24             4  away_win      0.500000   
10    2018020176            24             4  away_win      0.500000   
11    2016020164             8             4  home_win      0.555556   
12    2016020081             8             4  home_win      0.555556   
13    2017020965             8             4  away_win      0.555556   
14    2018020743             8             4  away_win      0.555556   
15    2018020934             8             4  home_win      0.555556   
16    2017021188            21             4  away_win      0.500000   
17    2016020444            21             4  away_win      0.500000   
18    2018020026            21             4  home_win      0.500000   
19    2017021183            25             4  home_win      0.500000   
20    2016020460            25             4  home_win      0.500000   
21    2018021235            25             4  home_win      0.500000   
22    2017020012            26             4  home_win      0.333333   
23    2016020016            26             4  away_win      0.333333   
24    2018020190            26             4  away_win      0.333333   
25    2017020850            27             4  away_win      0.500000   
26    2016020028            27             4  home_win      0.500000   
27    2018020214            27             4  away_win      0.500000   
28    2017020778             1             4  home_win      0.384615   
29    2016021180             1             4  home_win      0.384615   
...          ...           ...           ...       ...           ...   
3742  2016020685             6            16  away_win      0.333333   
3743  2017021048             6            16  home_win      0.333333   
3744  2018020863             6            16  home_win      0.333333   
3745  2016020433             3            16  away_win      0.666667   
3746  2017020615             3            16  away_win      0.666667   
3747  2018020728             3            16  home_win      0.666667   
3748  2017021239            19            16  away_win      0.571429   
3749  2016020562            19            16  home_win      0.571429   
3750  2017020093            19            16  home_win      0.571429   
3751  2016020467            19            16  away_win      0.571429   
3752  2016020195            19            16  away_win      0.571429   
3753  2018020024            19            16  away_win      0.571429   
3754  2018020155            19            16  home_win      0.571429   
3755  2017020328            14            16  home_win      0.000000   
3756  2016021124            14            16  home_win      0.000000   
3757  2018020341            14            16  home_win      0.000000   
3758  2017020300             5            16  away_win      1.000000   
3759  2016021137             5            16  away_win      1.000000   
3760  2018020650             5            16  away_win      1.000000   
3761  2017020874            31            16  home_win      0.000000   
3762  2017020135            31            16  home_win      0.000000   
3763  2018020438            31            16  home_win      0.000000   
3764  2016020281            22            16  home_win      0.444444   
3765  2016020827            22            16  awa

In [26]:
train = train.drop(columns=['team_id','score_first','shortName'],axis=1)
train

game_id  home_team_id  away_team_id   outcome  head_to_head  \
0     2016020045            16             4  home_win      0.666667   
1     2017020181            16             4  home_win      0.666667   
2     2018021142            16             4  away_win      0.666667   
3     2017020539             7             4  home_win      0.666667   
4     2016020973             7             4  away_win      0.666667   
5     2016020613             7             4  home_win      0.666667   
6     2018020318             7             4  home_win      0.666667   
7     2018020444             7             4  away_win      0.666667   
8     2016020561            24             4  home_win      0.500000   
9     2017020029            24             4  away_win      0.500000   
10    2018020176            24             4  away_win      0.500000   
11    2016020164             8             4  home_win      0.555556   
12    2016020081             8             4  home_win      0.555556   
13    2017020965             8             4  away_win      0.555556   
14    2018020743             8             4  away_win      0.555556   
15    2018020934             8             4  home_win      0.555556   
16    2017021188            21             4  away_win      0.500000   
17    2016020444            21             4  away_win      0.500000   
18    2018020026            21             4  home_win      0.500000   
19    2017021183            25             4  home_win      0.500000   
20    2016020460            25             4  home_win      0.500000   
21    2018021235            25             4  home_win      0.500000   
22    2017020012            26             4  home_win      0.333333   
23    2016020016            26             4  away_win      0.333333   
24    2018020190            26             4  away_win      0.333333   
25    2017020850            27             4  away_win      0.500000   
26    2016020028            27             4  home_win      0.500000   
27    2018020214            27             4  away_win      0.500000   
28    2017020778             1             4  home_win      0.384615   
29    2016021180             1             4  home_win      0.384615   
...          ...           ...           ...       ...           ...   
3742  2016020685             6            16  away_win      0.333333   
3743  2017021048             6            16  home_win      0.333333   
3744  2018020863             6            16  home_win      0.333333   
3745  2016020433             3            16  away_win      0.666667   
3746  2017020615             3            16  away_win      0.666667   
3747  2018020728             3            16  home_win      0.666667   
3748  2017021239            19            16  away_win      0.571429   
3749  2016020562            19            16  home_win      0.571429   
3750  2017020093            19            16  home_win      0.571429   
3751  2016020467            19            16  away_win      0.571429   
3752  2016020195            19            16  away_win      0.571429   
3753  2018020024            19            16  away_win      0.571429   
3754  2018020155            19            16  home_win      0.571429   
3755  2017020328            14            16  home_win      0.000000   
3756  2016021124            14            16  home_win      0.000000   
3757  2018020341            14            16  home_win      0.000000   
3758  2017020300             5            16  away_win      1.000000   
3759  2016021137             5            16  away_win      1.000000   
3760  2018020650             5            16  away_win      1.000000   
3761  2017020874            31            16  home_win      0.000000   
3762  2017020135            31            16  home_win      0.000000   
3763  2018020438            31            16  home_win      0.000000   
3764  2016020281            22            16  home_win      0.444444   
3765  2016020827            22            16  awa

In [27]:
train['%_score_first'] = (train['home_score_first'] / (train['away_score_first']+ train['home_score_first']))
train

game_id  home_team_id  away_team_id   outcome  head_to_head  \
0     2016020045            16             4  home_win      0.666667   
1     2017020181            16             4  home_win      0.666667   
2     2018021142            16             4  away_win      0.666667   
3     2017020539             7             4  home_win      0.666667   
4     2016020973             7             4  away_win      0.666667   
5     2016020613             7             4  home_win      0.666667   
6     2018020318             7             4  home_win      0.666667   
7     2018020444             7             4  away_win      0.666667   
8     2016020561            24             4  home_win      0.500000   
9     2017020029            24             4  away_win      0.500000   
10    2018020176            24             4  away_win      0.500000   
11    2016020164             8             4  home_win      0.555556   
12    2016020081             8             4  home_win      0.555556   
13    2017020965             8             4  away_win      0.555556   
14    2018020743             8             4  away_win      0.555556   
15    2018020934             8             4  home_win      0.555556   
16    2017021188            21             4  away_win      0.500000   
17    2016020444            21             4  away_win      0.500000   
18    2018020026            21             4  home_win      0.500000   
19    2017021183            25             4  home_win      0.500000   
20    2016020460            25             4  home_win      0.500000   
21    2018021235            25             4  home_win      0.500000   
22    2017020012            26             4  home_win      0.333333   
23    2016020016            26             4  away_win      0.333333   
24    2018020190            26             4  away_win      0.333333   
25    2017020850            27             4  away_win      0.500000   
26    2016020028            27             4  home_win      0.500000   
27    2018020214            27             4  away_win      0.500000   
28    2017020778             1             4  home_win      0.384615   
29    2016021180             1             4  home_win      0.384615   
...          ...           ...           ...       ...           ...   
3742  2016020685             6            16  away_win      0.333333   
3743  2017021048             6            16  home_win      0.333333   
3744  2018020863             6            16  home_win      0.333333   
3745  2016020433             3            16  away_win      0.666667   
3746  2017020615             3            16  away_win      0.666667   
3747  2018020728             3            16  home_win      0.666667   
3748  2017021239            19            16  away_win      0.571429   
3749  2016020562            19            16  home_win      0.571429   
3750  2017020093            19            16  home_win      0.571429   
3751  2016020467            19            16  away_win      0.571429   
3752  2016020195            19            16  away_win      0.571429   
3753  2018020024            19            16  away_win      0.571429   
3754  2018020155            19            16  home_win      0.571429   
3755  2017020328            14            16  home_win      0.000000   
3756  2016021124            14            16  home_win      0.000000   
3757  2018020341            14            16  home_win      0.000000   
3758  2017020300             5            16  away_win      1.000000   
3759  2016021137             5            16  away_win      1.000000   
3760  2018020650             5            16  away_win      1.000000   
3761  2017020874            31            16  home_win      0.000000   
3762  2017020135            31            16  home_win      0.000000   
3763  2018020438            31            16  home_win      0.000000   
3764  2016020281            22            16  home_win      0.444444   
3765  2016020827            22            16  awa

In [28]:

train['head_to_head'] = train['head_to_head']*100

In [61]:
train['%_score_first'] = train['%_score_first']

In [62]:
train

game_id  home_team_id  away_team_id   outcome  head_to_head  \
0     2016020045            16             4  home_win     66.666667   
1     2017020181            16             4  home_win     66.666667   
2     2018021142            16             4  away_win     66.666667   
3     2017020539             7             4  home_win     66.666667   
4     2016020973             7             4  away_win     66.666667   
5     2016020613             7             4  home_win     66.666667   
6     2018020318             7             4  home_win     66.666667   
7     2018020444             7             4  away_win     66.666667   
8     2016020561            24             4  home_win     50.000000   
9     2017020029            24             4  away_win     50.000000   
10    2018020176            24             4  away_win     50.000000   
11    2016020164             8             4  home_win     55.555556   
12    2016020081             8             4  home_win     55.555556   
13    2017020965             8             4  away_win     55.555556   
14    2018020743             8             4  away_win     55.555556   
15    2018020934             8             4  home_win     55.555556   
16    2017021188            21             4  away_win     50.000000   
17    2016020444            21             4  away_win     50.000000   
18    2018020026            21             4  home_win     50.000000   
19    2017021183            25             4  home_win     50.000000   
20    2016020460            25             4  home_win     50.000000   
21    2018021235            25             4  home_win     50.000000   
22    2017020012            26             4  home_win     33.333333   
23    2016020016            26             4  away_win     33.333333   
24    2018020190            26             4  away_win     33.333333   
25    2017020850            27             4  away_win     50.000000   
26    2016020028            27             4  home_win     50.000000   
27    2018020214            27             4  away_win     50.000000   
28    2017020778             1             4  home_win     38.461538   
29    2016021180             1             4  home_win     38.461538   
...          ...           ...           ...       ...           ...   
3742  2016020685             6            16  away_win     33.333333   
3743  2017021048             6            16  home_win     33.333333   
3744  2018020863             6            16  home_win     33.333333   
3745  2016020433             3            16  away_win     66.666667   
3746  2017020615             3            16  away_win     66.666667   
3747  2018020728             3            16  home_win     66.666667   
3748  2017021239            19            16  away_win     57.142857   
3749  2016020562            19            16  home_win     57.142857   
3750  2017020093            19            16  home_win     57.142857   
3751  2016020467            19            16  away_win     57.142857   
3752  2016020195            19            16  away_win     57.142857   
3753  2018020024            19            16  away_win     57.142857   
3754  2018020155            19            16  home_win     57.142857   
3755  2017020328            14            16  home_win      0.000000   
3756  2016021124            14            16  home_win      0.000000   
3757  2018020341            14            16  home_win      0.000000   
3758  2017020300             5            16  away_win    100.000000   
3759  2016021137             5            16  away_win    100.000000   
3760  2018020650             5            16  away_win    100.000000   
3761  2017020874            31            16  home_win      0.000000   
3762  2017020135            31            16  home_win      0.000000   
3763  2018020438            31            16  home_win      0.000000   
3764  2016020281            22            16  home_win     44.444444   
3765  2016020827            22            16  awa

In [63]:
from sklearn.linear_model import LogisticRegression  # for Logistic Regression algorithm
from sklearn.model_selection import train_test_split #to split the dataset for training and testing
from sklearn.neighbors import KNeighborsClassifier  # for K nearest neighbours
from sklearn import svm  #for Support Vector Machine (SVM) Algorithm
from sklearn import metrics #for checking the model accuracy
from sklearn.tree import DecisionTreeClassifier #for using Decision Tree Algoithm

In [64]:

train_1, test = train_test_split(train, test_size = 0.3)# in this our main data is split into train


In [65]:
train_X = train_1[['home_team_id','away_team_id','head_to_head','%_score_first']]
train_y = train_1.outcome

In [66]:
test_X=test[['home_team_id','away_team_id','head_to_head','%_score_first']]

In [67]:
test_y = test.outcome

In [68]:
train_X.head(5)

home_team_id  away_team_id  head_to_head  %_score_first
2877             2             1     33.333333       0.500000
196             28            19     55.555556       0.484962
34               1             4     38.461538       0.555556
2930            19             1      0.000000       0.522901
3245            12            27     50.000000       0.490040

In [69]:
test_X.head(5)

home_team_id  away_team_id  head_to_head  %_score_first
180             30            19     38.461538       0.454183
3456             8             5     66.666667       0.469697
985             16            30     38.461538       0.547619
2947            22             1     33.333333       0.481328
3315            22            27     38.461538       0.475410

In [70]:
train_y.head(5)

2877    home_win
196     away_win
34      away_win
2930    home_win
3245    away_win
Name: outcome, dtype: object

In [71]:
model = svm.SVC() #select the algorithm
model.fit(train_X,train_y) # we train the algorithm with the training data and the training output
prediction=model.predict(test_X) #now we pass the testing data to the trained algorithm
print('The accuracy of the SVM is:',metrics.accuracy_score(prediction,test_y))#now we check the accuracy of the algorithm. 
#we pass the predicted output by the model and the actual output

c:\users\simos\appdata\local\programs\python\python36-32\lib\site-packages\sklearn\svm\base.py:193: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


The accuracy of the SVM is: 0.5556537102473498


In [72]:
#Logistic regression
model = LogisticRegression()
model.fit(train_X,train_y)
prediction=model.predict(test_X)
print('The accuracy of the Logistic Regression is',metrics.accuracy_score(prediction,test_y))

The accuracy of the Logistic Regression is 0.6713780918727915


c:\users\simos\appdata\local\programs\python\python36-32\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


In [73]:
#Decision Tree
model=DecisionTreeClassifier()
model.fit(train_X,train_y)
prediction=model.predict(test_X)
print('The accuracy of the Decision Tree is',metrics.accuracy_score(prediction,test_y))

The accuracy of the Decision Tree is 0.5353356890459364


In [74]:
#KNN
model=KNeighborsClassifier(n_neighbors=3) #this examines 3 neighbours for putting the new data into a class
model.fit(train_X,train_y)
prediction=model.predict(test_X)
print('The accuracy of the KNN is',metrics.accuracy_score(prediction,test_y))

The accuracy of the KNN is 0.5556537102473498


In [60]:
#Therefore logistic regression is the best

### Any feedback is really appreciated. I am just dipping my toes in the water.